Update snakemake dependency to at least v9.8.1

HealthyPear opened the pull request:
## Overview

This PR successfully updates _showyourwork_ to **Snakemake v9.8.1** addressing breaking changes introduced by this new version while maintaining all existing functionality and performance characteristics.

It is to be noted that this is merely a "minimal" upgrade: it might be that somethings are not needed anymore on `showyourwork` side because of new features in Snakemake v9.8.1 or we are (yet) leveraging them. These points should be left for a future PR (or - even better - smaller, multiple ones).

To test/reproduce yourselves:

```shell
git clone https://github.com/showyourwork/showyourwork.git
cd showyourwork
gh pr checkout 556
conda env create -n showyourwork-snakemake9 python pip
conda activate showyourwork-snakemake9
pip install -e '.[dev]'
```

Closes #260 .

## 🔧 Changes Made

### Core Compatibility Updates

#### 1. **Async Architecture Adaptation** (`src/showyourwork/patches.py`)
- **Added `_run_async_safely()` helper function** to handle async coroutines from sync contexts
- **Updated `job_is_cached()`** to handle `cache.exists()` now being async
- **Updated `get_skippable_jobs()`** to handle `dag.file2jobs()` now being async  
- **Updated `patch_snakemake_cache_optimization()`** to handle `output.exists()` now being async

#### 2. **Dependency Management** (`pyproject.toml`)
- **Removed dependencies version pin**: now all dependencies are left unpinned, we can discuss if to pin them again later but my idea is that the development version should always be cutting-edge (even if this might increase maintenance)
- **Refactored new ruff configuration** new `ruff.toml` file

### Technical Implementation Details

#### Async Compatibility Strategy
The core challenge was that Snakemake v9.8.1 converted several key methods to async:
- `cache.exists(job, cache_mode)` → `await cache.exists(job, cache_mode)`
- `dag.file2jobs(file)` → `await dag.file2jobs(file)`
- `output.exists()` → `await output.exists()`

#### Solution: add a new `_run_async_safely()` Function

- runs an async coroutine safely, handling event loop context
- handles both running and non-running event loop scenarios
- uses `ThreadPoolExecutor` when needed to avoid blocking

### Modified Functions

#### `job_is_cached(job)` 
- **Before**: `cache_exists = cache.exists(job, cache_mode)`
- **After**: Uses `_run_async_safely(check_cache_exists())` with async wrapper

#### `get_skippable_jobs(dag)`
- **Before**: `file_jobs = dag.file2jobs(file)`  
- **After**: Uses `_run_async_safely(get_file_jobs())` with async wrapper

#### `patch_snakemake_cache_optimization(dag)`
- **Before**: `output_exists = output.exists()`
- **After**: Uses `_run_async_safely(check_output_exists())` with async wrapper

#### `patch_snakemake_wait_for_files`
- updated as upstream homonymous function changed

#### `patch_snakemake_logging()`
- added fallback handling for potential logger structure changes in Snakemake v9.8.1 which might for some reason not be tested in this project

#### `patch_snakemake_missing_input_leniency()`

- the logging interface in Snakemake has been greatly refactored to use the Python standard library and allow for plugins

### Additional Improvements



## 🧪 Testing

### Validation
- **All local tests are passing**: remote ones where already broken we will need to review them later
- **Test runtime**: consistent with previous runs
- **No regressions**: All previous functionality preserved
- **Clean implementation**: Removed all debug artifacts and temporary files

### Test Environment
- **Python**: 3.13.5
- **Snakemake**: v9.8.1

## 📋 Migration Notes

### Breaking Changes Addressed
1. **Async Methods**: All previously synchronous cache and DAG methods now require async handling
2. **Event Loop Management**: Proper handling of asyncio event loops in mixed sync/async environments  
3. **Version Compatibility**: Removed hard dependency on Snakemake 7.15.2

### Backward Compatibility
- **Cache optimization behavior unchanged**
- **API compatibility maintained** 
- **No user-facing changes required**
- **Existing workflows continue to work**, I ran both local tests and the quickstart tutorial (not yet the remote part!)


### Documentation

Nothing crazy for the moment, but I updated the CLI docs (`docs/cli.rst`) because now the Snakemake `--reason` flag is deprecated and always defaults to True.

HealthyPear committed:
Update pyproject.toml

- refactor Ruff options into their own config file
- update dependencies, leaving free their versions

HealthyPear committed:
Remove the 'reason' flag from  run_snakemake command.

From Snakemake 8 this option is deprecated and now always True by default.

HealthyPear committed:
snakemake.logging.logger is already a logger

HealthyPear committed:
Update patch_snakemake_missing_input_leniency

The logging framework of snakemake has been refactored to use the `logging` module from Python's standard library. The dag.py module of snakemake uses the logger object from there.

HealthyPear committed:
Update patch_snakemake_wait_for_files.

The original function in snakemake has been updated.

HealthyPear committed:
Update cahce optmization for snakemake 9.8.1

- added _run_async_safely for coroutine handling
- fixed async API calls
- enhanced error handling

HealthyPear committed:
Update patch_snakemake_logging for more robust backwards compatibility

HealthyPear commented:
The current errors in the CI are given by the outdated Python versions used there.

I will review them ASAP.

HealthyPear committed:
Set minimum Python version to 3.11 and update CI

HealthyPear committed:
Update docs build environment to use Python 3.11 and ensure compatibility with the latest packages.

HealthyPear commented:
@rodluger @dfm I do not have enough writing permissions to this project to access the repository settings

you should remove the check named "Run local integration tests (3.10, ubuntu-latest)" as required because given the changes in this PR it is now obsolete

we should set as required all the others (both local integration tests and unit tests for the selected python versions)

HealthyPear commented:
The documentation issue is given by the fact that the builder is getting Python 3.13 for which the issue is that global flags like (?s) (which enables dot-all mode, making . match newlines) must be at the very beginning of the regex pattern. The original pattern '"""((?s).*?)"""' had the flag inside the capture group, but it needs to be '(?s)"""(.*?)"""'.

I will try to fix it and to make it backwards compatible with Python 3.11 and 3.12 since these are the versions that should be supported by this PR.

HealthyPear committed:
Update documentation

- compatible with Python 3.11/12/13
- removed unsupported theme option 'logo_only'
- fixed docstring reference in workflow/scripts/arxiv.py to preprocess script

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -347,60 +371,118 @@ def patch_snakemake_wait_for_files():
-        force_stay_on_remote=False,
+        wait_for_local=False,
         ignore_pipe_or_service=False,
+        consider_local: set[_IOFile] = _CONSIDER_LOCAL_DEFAULT,
```
We are already requiring at least Python 3.11

HealthyPear committed:
Fix docstrin reference in src/showyourwork/workflow/rules/compile.smk to preprocess script

HealthyPear committed:
Create a fallback docs/projects.rst file to avoid Sphinx warnings in case of local builds

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -347,60 +371,118 @@ def patch_snakemake_wait_for_files():
         if missing:
-            get_logger().info(
-                f"Waiting at most {latency_wait} seconds for missing files."
+            fmt_missing = "\n".join
```
True, should be "\n".join(missing), as get_missing() returns the list of missing files...

HealthyPear committed:
Apply fixes spotted by Copilot during the upgrade to Snakemake 9.8.1

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -49,22 +50,34 @@ def patch_snakemake_missing_input_leniency():
+    from snakemake.logging import logger
+
+    # Store the original debug method
+    _original_debug = logger.debug
```
done

HealthyPear committed:
Make sure fmt_missing in patch_snakemake_wait_for_files has correct syntax

HealthyPear committed:
Update HISTORY

- fixed wrong link to v0.4.1 changelog
- added changelog for v0.4.2 and v0.4.3
- added changelog for development version

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

HealthyPear commented on pyproject.toml:
```diff
@@ -4,27 +4,30 @@ description = "A workflow for open-source scientific articles"
 ]
 readme = "README.md"
-requires-python = ">=3.8,<3.11"
+requires-python = ">=3.11"
```
Changelog has been updated accordingly.

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -347,60 +371,118 @@ def patch_snakemake_wait_for_files():
         if missing:
-            get_logger().info(
-                f"Waiting at most {latency_wait} seconds for missing files."
+            fmt_missing = "\n".join
```
I thought I did it before but I committed just one fix - now it has been fixed.

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -347,60 +371,118 @@ def patch_snakemake_wait_for_files():
-                time.sleep(1)
-            missing = "\n".join(get_missing())
+                time.sleep(sleep)
+            missing = fmt_missing(await get_missing(list_parent=True))
```
fixed

HealthyPear committed:
Fix calls to `fmt_missing` in patch_snakemake_wait_for_files

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -412,17 +494,22 @@ def job_is_cached(job):
 
     # Check if user requested caching for job
-    if not snakemake.workflow.workflow.is_cached_rule(job.rule):
+    cache_mode = snakemake.workflow.workflow.get_cache_mode(job.rule)
+    if cache_mode is None:
```
This suggestion is not relevant as the `is_cached_rule` method does not exist anymore in snakemake >7.15.2

HealthyPear commented on src/showyourwork/patches.py:
```diff
@@ -273,18 +286,29 @@ def patch_snakemake_logging():
         snakemake_logger.custom_stream_handler.setLevel(logging.ERROR)
         snakemake_logger.custom_stream_handler.setFormatter(SnakemakeFormatter())
-        snakemake_logger.logger.addHandler(snakemake_logger.custom_stream_handler)
+        snakemake_logger.addHandler(snakemake_logger.custom_stream_handler)
```
Also this is not relevant, as now in snakemake, `snakemake.logging.logger` (which is set at line 285) is already a logger so it does not have a further logger attribute (which was the case before)

dfm commented:
> you should remove the check named "Run local integration tests (3.10, ubuntu-latest)" as required because given the changes in this PR it is now obsolete

Done - thanks!

HealthyPear commented:
> > you should remove the check named "Run local integration tests (3.10, ubuntu-latest)" as required because given the changes in this PR it is now obsolete
> 
> Done - thanks!

Thank you!

Now the PR would be basically ready for merge, but I would appreciate if one of you main author guys checked/tested it.
Let me know what you think!

HealthyPear committed:
Update pyproject.toml for changelog

HealthyPear committed:
Add towncrier configuration

HealthyPear committed:
Update conf.py for changelog

HealthyPear committed:
Update HISTORY.rst for changelog

HealthyPear committed:
Add empty CHANGES.rst

HealthyPear committed:
Update changelog page

HealthyPear committed:
Add changelog Github workflow

HealthyPear committed:
Update contributors guide

HealthyPear committed:
Add docs/changes directory

HealthyPear committed:
Update docs conda environment

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

HealthyPear committed:
Update HISTORY.rst

add changelogs for v0.4.2 and v0.4.3.

HealthyPear committed:
Point changelog cross-ref to changelog.rst

HealthyPear committed:
Add news fragment for PR#557

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

dfm reviewed (approved):
LGTM. Thanks for working this out!

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.12.4 → v0.12.5](https://github.com/astral-sh/ruff-pre-commit/compare/v0.12.4...v0.12.5)

HealthyPear committed:
Merge branch 'main' into update-snakemake

HealthyPear committed:
Add news fragment for PR #556

HealthyPear committed:
fix typo in contributor guide

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

HealthyPear committed:
Update docs/environment.yml

sphinx-changelog is available only from PyPI

HealthyPear commented:
OK it seems that we are good to go!

Thank you for the review!
